### PR TITLE
[WIP, Pending on #10538]  Deprecate all EVP_MD_meth_ and EVP_CIPHER_meth_ functions in 3.0.0

### PR DESCRIPTION
**Pull request by levitte**
As a consequence, all engine functionality is deprecated as well,
since the `EVP_{TYPE}_meth_` functions form the basis for them.

### CODE REVIEW

**Comment by levitte** [comment]
There are several reason why this is currently a draft:

1.  its dependence on #9758 
2.  its bold nature, considering the team has quite divergent views on deprecation

File: crypto/evp/cmeth_lib.c [hunk 1]
```diff
@@ -14,6 +14,10 @@
 #include "internal/provider.h"
 #include "evp_locl.h"
 
+#if OPENSSL_API_3
```
Comment (slontis) [inline crypto/evp/cmeth_lib.c]: If this define is around the whole source file - then its a pity it cant be disabled via build.info

Comment (levitte) [reply crypto/evp/cmeth_lib.c]: Oh wait...  I did a modification in `Configure` for exactly this purpose (I added `$config{minapi}`).  So yeah, fix coming up!

**Comment by slontis** [comment]
Do some of the PKEY_methods also disappear?

**Comment by slontis** [comment]
I assume you have a cunning replacement plan that doesn't involve rewriting every engine as a provider in 3.0?

**Comment by levitte** [comment]
> I assume you have a cunning replacement plan that doesn't involve rewriting every engine as a provider in 3.0?

There's no reason why anyone should have to, really.

*Again, deprecation is not immediate removal, it's a warning that things will be removed some time in the future* (our release strategy says that deprecation must have been present for at least 5 years before removal)

The plan forward is that the engine mechanism will be removed entirely some time in the future, so people will *eventually* have to rewrite their engines as providers (the sooner the better in my opinion).  There's also been talks about making a bridge provider that can load engines with dynamic symbol magic (I have still not quite wrapped my brain around that), but that hasn't translated to code (yet?).

**Comment by levitte** [comment]
> Do some of the PKEY_methods also disappear?

They should.  I haven't started to look at them...

**Comment by levitte** [comment]
@petrovr, the rationale is basically that we're getting rid of the ENGINE interface and everything that has to do with that, in favor of provider.  We cannot, however, just *remove* everything, because we still want stability.  However, we must somehow warn people that this is going away in the future, and that's exactly what deprecation is for.

**Comment by petrovr** [comment]
No objection.

(Now I'm lost between what is deprecated and what is obsolete).

File: crypto/evp/cmeth_lib.c [hunk 2]
```diff
@@ -188,3 +190,4 @@ int (*EVP_CIPHER_meth_get_ctrl(const EVP_CIPHER *cipher))(EVP_CIPHER_CTX *,
     return cipher->ctrl;
 }
 
+#endif  /* OPENSSL_API_3 */
```
Comment (slontis) [inline crypto/evp/cmeth_lib.c]: Remove this line - the #ifdef was removed when you moved the exclusion into build.info

**Comment by levitte** [comment]
Someone else ended up doing this

**Pull request closed without merging**